engine/apps/shift_swaps/models/shift_swap_request.py

```diff
@@ -0,0 +1,140 @@
+import typing
+
+from django.conf import settings
+from django.core.validators import MinLengthValidator
+from django.db import models
+from django.utils import timezone
+
+from apps.shift_swaps import exceptions
+from common.public_primary_keys import generate_public_primary_key, increase_public_primary_key_length
+
+if typing.TYPE_CHECKING:
+    from apps.user_management.models import User
+
+
+def generate_public_primary_key_for_shift_swap_request() -> str:
+    prefix = "SSR"
+    new_public_primary_key = generate_public_primary_key(prefix)
+
+    failure_counter = 0
+    while ShiftSwapRequest.objects.filter(public_primary_key=new_public_primary_key).exists():
+        new_public_primary_key = increase_public_primary_key_length(
+            failure_counter=failure_counter, prefix=prefix, model_name="ShiftSwapRequest"
+        )
+        failure_counter += 1
+
+    return new_public_primary_key
+
+
+class ShiftSwapRequestQueryset(models.QuerySet):
+    def delete(self):
+        self.update(deleted_at=timezone.now())
+
+
+class ShiftSwapRequestManager(models.Manager):
+    def get_queryset(self):
+        return ShiftSwapRequestQueryset(self.model, using=self._db).filter(deleted_at=None)
+
+    def hard_delete(self):
+        return self.get_queryset().hard_delete()
+
+
+class ShiftSwapRequest(models.Model):
+
+    objects = ShiftSwapRequestManager()
+    objects_with_deleted = models.Manager()
+
+    public_primary_key = models.CharField(
+        max_length=20,
+        validators=[MinLengthValidator(settings.PUBLIC_PRIMARY_KEY_MIN_LENGTH + 1)],
+        unique=True,
+        default=generate_public_primary_key_for_shift_swap_request,
+    )
+
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+    deleted_at = models.DateTimeField(null=True)
+
+    schedule = models.ForeignKey(
+        to="schedules.OnCallSchedule", null=False, on_delete=models.CASCADE, related_name="shift_swap_requests"
+    )
+
+    time_zone = models.CharField(max_length=100)
+
+    swap_start = models.DateTimeField()
+    swap_end = models.DateTimeField()
```

Comment:
@matiasb based on your experience working w/ schedules/overrides. is there a benefit to storing datetimes as timezone aware/naive?

an alternative to what I did here is not have a `time_zone` column and simply have that as a required field when creating/updating objects, and we make `swap_start` and `swap_end` timezone aware before persisting them

Reply:
FWIW, since `USE_TZ` setting is set to `True` all datetime fields should be already timezone aware (using `UTC` timezone per the `TIME_ZONE` setting). But I think we don't really trust on this (I guess because the fields were not originally tz-aware?), and that's why there is a timezone field for shifts, for example. And then, we are saving a timestamp value without any tz conversion as an UTC-aware datetime working as a naive value, since if there is an associated timezone value, we assume the tz of that datetime value is based on that one, defaulting to `UTC` otherwise (see `convert_dt_to_schedule_timezone` in the `CustomOnCallShift` model). So we are in an intermediate scenario between using tz-aware and naive datetimes.

I think we should eventually migrate everything to fully use timezone aware datetimes because it should make things simpler/more readable when dealing with multiple TZs. Not really sure if introducing a different way to interpret datetime values will make things more complex or not (ie. we will need to know when to convert or not a datetime value?).

If I understand correctly, passing a tz-aware datetime here in a non-UTC tz should be converting the datetime value to UTC and store that in the DB, right? (so we will be working with converted to UTC datetime values later in our logic involving this).

Reply:
> I think we should eventually migrate everything to fully use timezone aware datetimes because it should make things simpler/more readable when dealing with multiple TZs

➕ ➕ 

> If I understand correctly, passing a tz-aware datetime here in a non-UTC tz should be converting the datetime value to UTC and store that in the DB, right

correct. it appears everything is persisted in the db in UTC. Django must handle the tz conversions when reading/writing:

<img width="341" alt="Screenshot 2023-07-20 at 19 38 15" src="https://github.com/grafana/oncall/assets/9406895/532583e3-bdd7-4aaf-bf5f-23b881d814e8">


```python3
>>> from apps.shift_swaps.models import ShiftSwapRequest
>>> from apps.user_management.models import User
>>> from apps.schedules.models import OnCallSchedule
>>> u = User.objects.first()
>>> s = OnCallSchedule.objects.first()
>>> ssr_utc = ShiftSwapRequest.objects.create(beneficiary=u, schedule=s, time_zone="Europe/Luxembourg", swap_start="2023-07-20T23:00:00Z", swap_end="2023-07-20T23:30:00Z")
>>> ssr_lux = ShiftSwapRequest.objects.create(beneficiary=u, schedule=s, time_zone="Europe/Luxembourg", swap_start="2023-07-20T23:00:00-04:00", swap_end="2023-07-20T23:30:00-04:00")
>>> ssr = ShiftSwapRequest.objects.create(beneficiary=u, schedule=s, time_zone="Europe/Luxembourg", swap_start="2023-07-20 23:00:00", swap_end="2023-07-20 23:30:00")
>>> ssr_utc.swap_end
'2023-07-20T23:30:00Z'
>>> ssr.swap_end
'2023-07-20 23:30:00'
>>> ssr_lux.swap_end
'2023-07-20T23:30:00-04:00'
```

Reply:
actually, come to think of it, maybe we can get rid of this column here and just use the schedule's `time_zone`, since a shift swap will always have a schedule related to it. wdyt?

Reply:
FYI, `time_zone` information is not "persisted":

```
>>> ssr_lux.swap_end
'2023-07-20T23:30:00-04:00'
>>> ssr_lux.refresh_from_db()
>>> ssr_lux.swap_end
datetime.datetime(2023, 7, 21, 3, 30, tzinfo=<UTC>)
```

everything is stored in `UTC` in the DB, and next time you fetch from DB, you don't know which was the original tz/datetime value that was provided (which is not necessarily an issue, just for clarification).

> actually, come to think of it, since a shift swap will always have a schedule related to it, we can probably get rid of this column, and just use the schedule's time_zone (maybe add a time_zone @property on this model for easier access).

Hm... I don't think we can use the schedule TZ for this because (as things work today for schedules/shifts, as mentioned in the previous comment) that would mean the datetime you get from the DB was specified using that TZ (which may not be the case?), and directly setting that TZ as the `swap_start` or `swap_end` values will potentially give unexpected results.

So... I think we could keep this information as UTC-aware datetimes (Django will do the right conversion if a TZ-aware datetime is given), and work with these fields always knowing/expecting UTC timestamps, or we could keep the existing approach where if a TZ is provided, save that value in the model and use datetime values as naive ones (this is the intermediate "awareness" scenario that we have I was referring to).

I think I would prefer the first option in this case (unless there are any objections about moving to a full Django tz aware datetimes gradually), although it diverges from what we have elsewhere (and we lose information about the original TZ in which the datetimes were defined).

Makes sense?

Reply:
@matiasb see [this commit](https://github.com/grafana/oncall/pull/2597/commits/399ede4767bcc4fdab1f9f1308bb8735643ca8af). 

based on our discussion, clients will be _forced_ to specify timezone when creating/updating `swap_start` and/or `swap_end`. I used ISO-8601 as a starting point, which allows specifying time zone as either UTC (ie. `Z` at the end) or UTC offset (ex. `+00:00` or `-07:00). However, ISO-8601 also allows not specifying tz (ex. 2023-07-20 12:00:00), I got rid of this.

There's a new custom DRF serializer field, [`TimeZoneAwareDatetimeField`](https://github.com/grafana/oncall/pull/2597/files#diff-27f90cb3c156840972f72d16e217d9808cd80ec2f7783188100d33d534040abbR167-R206), which handles all the input validation + formatting for us.

It might be best to check out the test cases:
- [serializer field test cases](https://github.com/grafana/oncall/pull/2597/files#diff-9f7a24d79d62efdf8bf51c93c9ca8a2074f1ac4f019f9e8fb1b46fe6477008bbR64-R102)
- API related test cases ([create](https://github.com/grafana/oncall/pull/2597/files#diff-de6c9e0686c9bd63dabc4b2824a0f2a381d799ad35add03705892b1ddcdce67fR187-R242) and [update](https://github.com/grafana/oncall/pull/2597/files#diff-de6c9e0686c9bd63dabc4b2824a0f2a381d799ad35add03705892b1ddcdce67fR307-R359))

let me know what you think.